File: kirun-js/src/engine/function/system/date/IsBetween.ts
import { Schema } from '../../../json/schema/Schema';
import { EventResult } from '../../../model/EventResult';
import { FunctionOutput } from '../../../model/FunctionOutput';
import { Parameter } from '../../../model/Parameter';
import { Namespaces } from '../../../namespaces/Namespaces';
import { FunctionExecutionParameters } from '../../../runtime/FunctionExecutionParameters';
import { MapUtil } from '../../../util/MapUtil';
import { AbstractDateFunction } from './AbstractDateFunction';
import { getDateTime } from './common';

export class IsBetween extends AbstractDateFunction {
    public static readonly PARAMETER_START_TIMESTAMP_NAME = 'startTimestamp';
    public static readonly PARAMETER_END_TIMESTAMP_NAME = 'endTimestamp';

    public static readonly PARAMETER_CHECK_TIMESTAMP_NAME = 'checkTimestamp';

    constructor() {
        super(
            'IsBetween',
            IsBetween.EVENT_BOOLEAN,
            Parameter.of(
                IsBetween.PARAMETER_START_TIMESTAMP_NAME,
                Schema.ofRef(Namespaces.DATE + '.Timestamp'),
            ),
            Parameter.of(
                IsBetween.PARAMETER_END_TIMESTAMP_NAME,
                Schema.ofRef(Namespaces.DATE + '.Timestamp'),
            ),
            Parameter.of(
                IsBetween.PARAMETER_CHECK_TIMESTAMP_NAME,
                Schema.ofRef(Namespaces.DATE + '.Timestamp'),
            ),
        );
    }

    protected async internalExecute(context: FunctionExecutionParameters): Promise<FunctionOutput> {
        const startTimestamp = context
            .getArguments()
            ?.get(IsBetween.PARAMETER_START_TIMESTAMP_NAME);
        const endTimestamp = context.getArguments()?.get(IsBetween.PARAMETER_END_TIMESTAMP_NAME);
        const checkTimestamp = context
            .getArguments()
            ?.get(IsBetween.PARAMETER_CHECK_TIMESTAMP_NAME);

        let startDateTime = getDateTime(startTimestamp);
        let endDateTime = getDateTime(endTimestamp);
        const checkDateTime = getDateTime(checkTimestamp);

        if (startDateTime > endDateTime) {
            [startDateTime, endDateTime] = [endDateTime, startDateTime];
        }

        return new FunctionOutput([
            EventResult.outputOf(
                MapUtil.of(
                    IsBetween.EVENT_RESULT_NAME,
                    startDateTime <= checkDateTime && checkDateTime <= endDateTime,
                ),
            ),
        ]);
    }
}

File: kirun-py/src/kirun_py/function/system/date/is_between.py
from __future__ import annotations

from typing import TYPE_CHECKING

from kirun_py.json.schema.schema import Schema
from kirun_py.model.event_result import EventResult
from kirun_py.model.function_output import FunctionOutput
from kirun_py.model.parameter import Parameter
from kirun_py.namespaces.namespaces import Namespaces
from kirun_py.function.system.date.abstract_date_function import AbstractDateFunction
from kirun_py.function.system.date.common import get_datetime

if TYPE_CHECKING:
    from kirun_py.runtime.function_execution_parameters import FunctionExecutionParameters


class IsBetween(AbstractDateFunction):

    PARAMETER_START_TIMESTAMP_NAME: str = 'startTimestamp'
    PARAMETER_END_TIMESTAMP_NAME: str = 'endTimestamp'
    PARAMETER_CHECK_TIMESTAMP_NAME: str = 'checkTimestamp'

    def __init__(self) -> None:
        super().__init__(
            'IsBetween',
            AbstractDateFunction.EVENT_BOOLEAN,
            Parameter.of(
                IsBetween.PARAMETER_START_TIMESTAMP_NAME,
                Schema.of_ref(Namespaces.DATE + '.Timestamp'),
            ),
            Parameter.of(
                IsBetween.PARAMETER_END_TIMESTAMP_NAME,
                Schema.of_ref(Namespaces.DATE + '.Timestamp'),
            ),
            Parameter.of(
                IsBetween.PARAMETER_CHECK_TIMESTAMP_NAME,
                Schema.of_ref(Namespaces.DATE + '.Timestamp'),
            ),
        )

    async def internal_execute(self, context: FunctionExecutionParameters) -> FunctionOutput:
        args = context.get_arguments() or {}

        start_dt = get_datetime(args.get(IsBetween.PARAMETER_START_TIMESTAMP_NAME))
        end_dt = get_datetime(args.get(IsBetween.PARAMETER_END_TIMESTAMP_NAME))
        check_dt = get_datetime(args.get(IsBetween.PARAMETER_CHECK_TIMESTAMP_NAME))

        if start_dt > end_dt:
            start_dt, end_dt = end_dt, start_dt

        result = start_dt <= check_dt <= end_dt

        return FunctionOutput([
            EventResult.output_of(
                {AbstractDateFunction.EVENT_RESULT_NAME: result}
            ),
        ])
